lib/services/unified_resource.go

```diff
@@ -0,0 +1,475 @@
+		c.tree.ReplaceOrInsert(&Item{Key: keyOf(resource), Value: resource})
+	}
+	return nil
+}
+
+// Delete deletes item by key, returns NotFound error
+// if item does not exist
+func (c *UnifiedResourceCache) delete(key []byte) error {
+	if len(key) == 0 {
+		return trace.BadParameter("missing parameter key")
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, ok := c.tree.Delete(&Item{Key: key}); !ok {
+		return trace.NotFound("key %q is not found", string(key))
+	}
+	return nil
+}
+
+func (c *UnifiedResourceCache) getRange(startKey, endKey []byte, limit int) ([]resource, error) {
+	if len(startKey) == 0 {
+		return nil, trace.BadParameter("missing parameter startKey")
+	}
+	if len(endKey) == 0 {
+		return nil, trace.BadParameter("missing parameter endKey")
+	}
+	if limit <= 0 {
+		limit = backend.DefaultRangeLimit
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	var res []resource
+	c.tree.AscendRange(&Item{Key: startKey}, &Item{Key: endKey}, func(item *Item) bool {
+		res = append(res, item.Value)
+		if limit > 0 && len(res) >= limit {
+			return false
+		}
+		return true
+	})
+
+	if len(res) == backend.DefaultRangeLimit {
+		c.log.Warnf("Range query hit backend limit. (this is a bug!) startKey=%q,limit=%d", startKey, backend.DefaultRangeLimit)
+	}
+	return res, nil
+}
+
+// GetUnifiedResources returns a list of all resources stored in the current unifiedResourceCollector tree
+func (c *UnifiedResourceCache) GetUnifiedResources(ctx context.Context) ([]types.ResourceWithLabels, error) {
+	if err := c.refreshStaleResources(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	result, err := c.getRange(backend.Key(prefix), backend.RangeEnd(backend.Key(prefix)), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err, "getting unified resource range")
+	}
+
+	resources := make([]types.ResourceWithLabels, 0, len(result))
+	for _, item := range result {
+		resources = append(resources, item.CloneResource())
+	}
+
+	return resources, nil
+}
+
+type ResourceGetter interface {
+	NodesGetter
+	DatabaseServersGetter
+	AppServersGetter
+	WindowsDesktopGetter
+	KubernetesServerGetter
+	SAMLIdpServiceProviderGetter
+}
+
+// newWatcher starts and returns a new resource watcher for unified resources.
+func newWatcher(ctx context.Context, resourceCache *UnifiedResourceCache, cfg ResourceWatcherConfig) error {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err, "setting defaults for unified resource watcher config")
+	}
+
+	if _, err := newResourceWatcher(ctx, resourceCache, cfg); err != nil {
+		return trace.Wrap(err, "creating a new unified resource watcher")
+	}
+	return nil
+}
+
+func keyOf(resource types.Resource) []byte {
+	switch r := resource.(type) {
+	case types.Server:
+		return backend.Key(prefix, r.GetHostname(), r.GetName(), r.GetKind())
+	default:
+		return backend.Key(prefix, r.GetName(), r.GetKind())
+	}
+}
+
+func (c *UnifiedResourceCache) getResourcesAndUpdateCurrent(ctx context.Context) error {
```

Comment:
I did the one big insert but the function seems a bit ugly because I have to append to my `newResources` slice every time i run a new getter. Is there a cleaner way to do it? 

As of now, the lock should only be acquired three times. one to `clear`, one for the big `putResources` and once for updating `c.stale`. 

Also, as spoke of in DMs but putting here for posterity, this will prevent "halfstates" where say we were able to get and update nodes, but not the other resources. we wouldn't want a tree only sorta updating (even if it does stay stale).